go/vt/vttablet/tabletmanager/tm_state.go

```diff
@@ -0,0 +1,392 @@
+				ts.blacklistedTables[tc.TabletType] = tc.BlacklistedTables
+			}
+		}
+	}
+
+	if srvKeyspace != nil {
+		ts.tabletControls = make(map[topodatapb.TabletType]bool)
+		for _, partition := range srvKeyspace.GetPartitions() {
+			for _, tabletControl := range partition.GetShardTabletControls() {
+				if key.KeyRangeEqual(tabletControl.GetKeyRange(), ts.KeyRange()) {
+					if tabletControl.QueryServiceDisabled {
+						ts.tabletControls[partition.GetServedType()] = true
+					}
+					break
+				}
+			}
+		}
+	}
+
+	ts.updateLocked(ctx)
+}
+
+func (ts *tmState) ChangeTabletType(ctx context.Context, tabletType topodatapb.TabletType) error {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	log.Infof("Changing Tablet Type: %v", tabletType)
+
+	if tabletType == topodatapb.TabletType_MASTER {
+		masterTermStartTime := logutil.TimeToProto(time.Now())
+
+		// Update the tablet record first.
+		_, err := topotools.ChangeType(ctx, ts.tm.TopoServer, ts.tm.tabletAlias, tabletType, masterTermStartTime)
+		if err != nil {
+			return err
+		}
+
+		ts.tablet.Type = tabletType
+		ts.tablet.MasterTermStartTime = masterTermStartTime
+	} else {
+		ts.tablet.Type = tabletType
+		ts.tablet.MasterTermStartTime = nil
+	}
+
+	s := topoproto.TabletTypeLString(tabletType)
+	statsTabletType.Set(s)
+	statsTabletTypeCount.Add(s, 1)
+
+	ts.updateLocked(ctx)
+	ts.publishStateLocked(ctx)
+	ts.tm.notifyShardSync()
+	return nil
+}
+
+func (ts *tmState) SetMysqlPort(mport int32) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	ts.tablet.MysqlPort = mport
+	ts.publishStateLocked(ts.ctx)
+}
+
+// UpdateTablet must be called during initialization only.
+func (ts *tmState) UpdateTablet(update func(tablet *topodatapb.Tablet)) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	update(ts.tablet)
+}
+
+func (ts *tmState) updateLocked(ctx context.Context) {
+	span, ctx := trace.NewSpan(ctx, "tmState.update")
+	defer span.Finish()
+
+	if !ts.isOpen {
+		return
+	}
+
+	terTime := logutil.ProtoToTime(ts.tablet.MasterTermStartTime)
+	ts.publishForDisplay()
+
+	// Disable TabletServer first so the nonserving state gets advertised
+	// before other services are shutdown.
+	reason := ts.canServe(ts.tablet.Type)
+	if reason != "" {
+		log.Infof("Disabling query service: %v", reason)
+		if err := ts.tm.QueryServiceControl.SetServingType(ts.tablet.Type, terTime, false, reason); err != nil {
+			log.Errorf("SetServingType(serving=false) failed: %v", err)
+		}
+	}
+
+	if err := ts.applyBlacklist(ctx); err != nil {
+		log.Errorf("Cannot update blacklisted tables rule: %v", err)
+	}
+
+	ts.tm.replManager.SetTabletType(ts.tablet.Type)
+
+	if ts.tm.UpdateStream != nil {
+		if topo.IsRunningUpdateStream(ts.tablet.Type) {
+			ts.tm.UpdateStream.Enable()
+		} else {
+			ts.tm.UpdateStream.Disable()
+		}
+	}
+
+	if ts.tm.VREngine != nil {
+		if ts.tablet.Type == topodatapb.TabletType_MASTER {
+			ts.tm.VREngine.Open(ts.tm.BatchCtx)
+		} else {
+			ts.tm.VREngine.Close()
+		}
+	}
```

Comment:
UpdateStream and VREngine are not set in the test. See if can be added in test tm initialize

Reply:
The unit test framework doesn't support these. But there are many end2end tests for this.